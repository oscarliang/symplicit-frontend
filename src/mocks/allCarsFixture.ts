export default {
  '1701088': {
    _id: '656083b26a2462000a137319',
    id: 1701088,
    brand: 'PORSCHE',
    name: 'test1',
    price: '34324',
    drive: '2wd|4wd|awd|',
    imageUrl:
      'https://images.carexpert.com.au/app/uploads/2023/09/BMW-i7-M70-xDrive_HERO-16x9-2.jpg',
  },
  '1703800': {
    _id: '65612c1df6050feba2d25132',
    id: 1703800,
    brand: 'PORSCHE',
    name: 'test2',
    price: '4545',
    drive: '2wd|4wd|awd|',
    imageUrl:
      'https://editorial.pxcrush.net/carsales/general/editorial/porsche-911-gts-0182.jpg?width=1024&height=683',
  },
  '1704417': {
    _id: '656131f7f6050feba2d25147',
    id: 1704417,
    brand: 'BMW',
    name: 'test3',
    price: '999921212',
    drive: '2wd|4wd|',
    imageUrl:
      'https://images.carexpert.com.au/app/uploads/2023/09/BMW-i7-M70-xDrive_HERO-16x9-2.jpg',
  },
};
