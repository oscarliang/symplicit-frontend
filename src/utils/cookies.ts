// Based on `js-cookie`

export const ACCESS_TOKEN_KEY = 'accessToken';

export function getCookie(key: string): null | string {
  const cookies = document.cookie ? document.cookie.split('; ') : [];
  for (let i = 0; i < cookies.length; i += 1) {
    const parts = cookies[i].split('=');
    let value = parts.slice(1).join('=');

    if (value[0] === '"') {
      value = value.slice(1, -1);
    }

    const foundKey = decodeURIComponent(parts[0]);

    if (key === foundKey) {
      return value.replace(/(%[\dA-F]{2})+/gi, decodeURIComponent);
    }
  }

  return null;
}

interface SetCookieOptions {
  cookieName: string;
  cookiePath?: string;
  cookieValue: string;
  expireSeconds?: number;
}

export function setCookie({
  cookieName,
  cookiePath,
  cookieValue,
  expireSeconds,
}: SetCookieOptions): void {
  let cookieString = `${cookieName}=${cookieValue}`;
  if (expireSeconds) {
    // eslint-disable-next-line rulesdir/prefer-use-date
    const today = new Date();
    // eslint-disable-next-line rulesdir/prefer-use-date
    const expire = new Date();
    expire.setTime(today.getTime() + expireSeconds * 1000);
    cookieString += `;expires=${expire.toUTCString()}`;
  }
  if (cookiePath) {
    cookieString += `;path=${cookiePath}`;
  }
  document.cookie = cookieString;
}

export function removeCookie(cookieName: string): void {
  // eslint-disable-next-line max-len
  document.cookie = `${cookieName}=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;`;
}
