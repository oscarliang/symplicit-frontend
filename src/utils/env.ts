// eslint-disable-next-line import/prefer-default-export
export const ENV = {
  backendHost: 'oscar.db.docker', // used in ssr api calling in docker
  cookiePrefix: 'symplicit',
  debugger: false,
  graphqlHost: 'localhost',
  graphqlPort: 3003,
  restApiPort: 3002,
  port: 4000,
};
