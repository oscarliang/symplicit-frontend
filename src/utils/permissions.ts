import { PermissionType } from '../types/permission';

// eslint-disable-next-line import/prefer-default-export
export function hasPermission(
  permissions: Record<string, number>,
  module: string,
  permissionTypes: PermissionType[],
): boolean {
  if (!(module in permissions)) {
    return false;
  }

  let granted = false;
  for (let i = 0; i < permissionTypes.length; i++) {
    // eslint-disable-next-line no-bitwise
    if (permissions[module] & permissionTypes[i]) {
      granted = true;
      break;
    }
  }
  return granted;
}
