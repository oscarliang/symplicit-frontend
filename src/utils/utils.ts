export function isObject(object: unknown): object is Record<string, unknown> {
  return !!object && !Array.isArray(object) && typeof object === 'object';
}

/* eslint-disable no-param-reassign */
export function merge(...objects: unknown[]): unknown {
  return objects.reduce((current: Record<string, unknown>, object) => {
    if (!isObject(object)) return current;

    Object.keys(object).forEach((key) => {
      const currentValue = current[key];
      const objectValue = object[key];

      if (isObject(currentValue) && isObject(objectValue)) {
        current[key] = merge(currentValue, objectValue);
      } else if (Array.isArray(objectValue)) {
        current[key] = objectValue.slice(0);
      } else {
        current[key] = objectValue;
      }
    });

    return current;
  }, {});
}
/* eslint-enable no-param-reassign */

export function cssStylesToObject(styles: string): Record<string, string> {
  const reactStyles: Record<string, string> = {};
  styles.split(';').forEach((style) => {
    const [key, value] = style.split(':').map((s) => (s ? s.trim() : s));
    if (!key || !value) return;
    const camelcaseKey = key.replace(/(-\w)/g, (word: string) =>
      word[1].toUpperCase(),
    );
    reactStyles[camelcaseKey] = value;
  });
  return reactStyles;
}

export function snakeToCamel(value: string): string {
  return value
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .split(/_/g)
    .map((segment, idx) => {
      const firstLetter = segment.slice(0, 1);
      const rest = segment.slice(1);
      return `${firstLetter[
        idx === 0 ? 'toLowerCase' : 'toUpperCase'
      ]()}${rest.toLowerCase()}`;
    })
    .join('');
}

/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

const excludedKeys = ['_id'];

export function keysToCamel(obj: any): any {
  if (!obj || typeof obj !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {
    return obj.map((item) => keysToCamel(item));
  }

  const json: Record<string, any> = obj as Record<string, any>;
  const newObj: Record<string, any> = {};
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  Object.keys(obj).forEach((key) => {
    if (excludedKeys.includes(key)) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      newObj[key] = json[key];
    } else {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      newObj[snakeToCamel(key)] = keysToCamel(json[key]);
    }
  });

  return newObj;
}
/* eslint-enable @typescript-eslint/no-unsafe-return */
/* eslint-enable @typescript-eslint/no-explicit-any */

export function camelToSnake(string: string, splitOnNumbers = true): string {
  return string.replace(
    splitOnNumbers ? /[A-Z0-9]/g : /[A-Z]/g,
    (letter) => `_${letter.toLowerCase()}`,
  );
}

export function camelToWords(string: string): string {
  return string.replace(/[A-Z]/g, (letter) => ` ${letter.toLowerCase()}`);
}

export function wordToCamel(value: string): string {
  return value
    .trim()
    .split(/ /g)
    .map((segment, idx) => {
      const firstLetter = segment.slice(0, 1);
      const rest = segment.slice(1);
      return `${firstLetter[
        idx === 0 ? 'toLowerCase' : 'toUpperCase'
      ]()}${rest.toLowerCase()}`;
    })
    .join('');
}

export function camelKeysToSnake(
  obj: Record<string, unknown>,
  splitOnNumbers = true,
): Record<string, unknown> {
  return Object.entries(obj).reduce(
    (newObj, [key, value]) => ({
      ...newObj,
      [camelToSnake(key, splitOnNumbers)]: value,
    }),
    {},
  );
}

export function nameToInitials(name: string): string {
  // Basic function to create initials from a name.
  // - Not smart to detect if index = 1 is middlename or lastname
  // - If name is one word, it uses two first letters of that single word
  const nameParts = name.replace(/\t/g, ' ').trim().split(' ');
  return nameParts.length === 1
    ? nameParts[0].split('').slice(0, 2).join('') || ''
    : `${nameParts[0][0]}${nameParts.slice(-1)[0][0]}`;
}

function titleCaseWord(word: string): string {
  if (word.toLowerCase() === 's ') {
    // Assumes this is the single "s" after an "'s",
    // so want to keep it lowercase.
    return word.toLowerCase();
  }
  return word.charAt(0).toUpperCase() + word.substring(1).toLowerCase();
}

export function titleCase(name: string | null): string {
  if (!name) return '';

  // Transform to title case
  const basicCapitalize = name
    .split(' ')
    .map(
      (part) => `${part.charAt(0).toUpperCase()}${part.slice(1).toLowerCase()}`,
    )
    .join(' ');

  return basicCapitalize
    .replace(/[^\s\-']+[\s\-']*/g, (word: string) => titleCaseWord(word))
    .replace(/\b(Van|De|Der|Da|Von)\b/g, (nobiliaryParticle) =>
      nobiliaryParticle.toLowerCase(),
    )
    .replace(/\b(And|To|The|From|Of|At|For|By|On|With|In)\b/g, (specialCase) =>
      // Some special cases:
      // - `TO`, `THE` and `EDITOR` LETTERS TO THE EDITOR
      // - `AND` in case `AUTHOR1 AND AUTHOR2`
      specialCase.toLowerCase(),
    )
    .replace(
      /Mc(.)/g,
      (_match, letter3: string) => `Mc${letter3.toUpperCase()}`,
    )
    .replace(/^(\w{1})/g, (firstLetter: string) => firstLetter.toUpperCase());
}
