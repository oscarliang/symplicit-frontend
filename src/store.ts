import { PayloadAction, configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { loadingBarReducer } from 'react-redux-loading-bar';
import { combineReducers } from 'redux';
import * as Redux from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';

import carsReducer from './ducks/cars';
import globalReducer from './ducks/global';
import LoginReducer, { LOG_OUT_ACTION } from './ducks/login';
import permissionReducer from './ducks/permission';
import { merge } from './utils/utils';

const ACTION_HYDRATE = 'HYDRATE';
export function hydrateAction(
  state: Partial<RootState>,
): PayloadAction<Partial<RootState>> {
  return {
    payload: state as RootState,
    type: ACTION_HYDRATE,
  };
}

const combinedReducer = combineReducers({
  authenticated: LoginReducer,
  cars: carsReducer,
  global: globalReducer,
  loadingBar: loadingBarReducer,
  permissions: permissionReducer,
});

export function createStore(configure?: (state: RootState) => RootState) {
  const middleware: [Redux.Middleware] = [thunk];

  if (
    process.env.NODE_ENV !== 'production' &&
    process.env.NEXT_PUBLIC_DEBUGGER
  ) {
    // add the previous state and next state info
    middleware.push(
      createLogger({
        collapsed: true,
        level: 'info',
      }),
    );
  }

  const store = configureStore({
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(middleware),
    reducer: (
      state: RootState | undefined,
      action: PayloadAction<Partial<RootState>>,
    ) => {
      // Clear all data in redux store to initial.
      if (action.type === LOG_OUT_ACTION) {
        // eslint-disable-next-line no-param-reassign
        state = undefined;
      }

      if (action.type === ACTION_HYDRATE) {
        return merge(state, action.payload) as RootState;
      }
      return combinedReducer(state, action);
    },
  });

  if (configure) {
    store.dispatch(hydrateAction(configure(store.getState())));
  }

  return store;
}

export const store = createStore();

export type RootState = ReturnType<typeof combinedReducer>;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
