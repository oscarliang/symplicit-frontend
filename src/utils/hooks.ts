import { useEffect, useState } from 'react';

import { useAppSelector } from '../store';

import { ACCESS_TOKEN_KEY, getCookie, setCookie } from './cookies';

export default function useToken() {
  const { isAuthenticated } = useAppSelector((state) => state.authenticated);
  const getToken = (): string | null => {
    // eslint-disable-next-line rulesdir/no-typeof-window-outside-useeffect
    if (typeof window === 'undefined') return null;

    const tokenString = getCookie(`${ACCESS_TOKEN_KEY}`);
    if (!tokenString) {
      return null;
    }

    return tokenString;
  };

  const [token, setToken] = useState(getToken());

  // have to use this solution to trigger refresh the store state. Full.tsx - getToken need the refreshed token
  useEffect(() => {
    setToken(getCookie(`${ACCESS_TOKEN_KEY}`));
  }, [isAuthenticated]);

  const saveToken = (userToken: string) => {
    setCookie({
      cookieName: `${ACCESS_TOKEN_KEY}`,
      cookiePath: '/',
      cookieValue: userToken,
      expireSeconds: 7 * 24 * 60 * 60,
    });
  };

  return {
    setToken: saveToken,
    token,
  };
}
