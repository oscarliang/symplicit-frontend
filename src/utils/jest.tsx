import { EnhancedStore } from '@reduxjs/toolkit';
import * as React from 'react';
import { Provider } from 'react-redux';
import { SWRConfig } from 'swr';

import { RootState } from 'store';

interface TestWrapperProps {
  children: React.ReactNode;
  store: EnhancedStore<RootState>;
}

// eslint-disable-next-line import/prefer-default-export
export function TestWrapper({
  children,
  store,
}: TestWrapperProps): React.ReactElement {
  return (
    <SWRConfig value={{ provider: () => new Map() }}>
      <Provider store={store}>{children}</Provider>
    </SWRConfig>
  );
}
