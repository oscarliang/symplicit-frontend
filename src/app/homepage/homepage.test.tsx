import { EnhancedStore } from '@reduxjs/toolkit';
import { fireEvent, render, screen } from '@testing-library/react';
import * as React from 'react';

import { AllCar, RoleModules } from 'mocks/homepageFixture';
import AllCarsFixture from 'mocks/allCarsFixture';
import { RootState, createStore } from 'store';
import { TestWrapper } from 'utils/jest';

import HomepageComponent from './component';

describe('<HomepageContainer />', () => {
  let store: EnhancedStore<RootState>;

  beforeEach(() => {
    store = createStore((state: RootState) => ({
      ...state,
      cars: AllCarsFixture,
      global: {
        isClientSide: true,
        message: 'test',
        selectedCar: {
          brand: '',
          drive: '',
          imageUrl: '',
          name: '',
          price: '',
          id: 0,
          _id: '',
        },
      },
    }));
  });

  it('renders', () => {
    expect.assertions(1);

    const { container } = render(
      <TestWrapper store={store}>
        <HomepageComponent
          allCars={AllCar}
          roleModules={RoleModules as Record<string, number>}
        />
      </TestWrapper>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('button', () => {
    expect.assertions(4);

    render(
      <TestWrapper store={store}>
        <HomepageComponent
          allCars={AllCar}
          roleModules={RoleModules as Record<string, number>}
        />
      </TestWrapper>,
    );

    expect(screen.getByTestId('add-car-btn')).not.toHaveClass('disabled');
    expect(screen.getByTestId('reset-car-btn')).not.toHaveClass('disabled');
    expect(screen.getByTestId('update-car-btn')).toHaveClass('disabled');

    fireEvent.click(screen.getByAltText('test1'));
    expect(screen.getByTestId('update-car-btn')).not.toHaveClass('disabled');
  });
});
