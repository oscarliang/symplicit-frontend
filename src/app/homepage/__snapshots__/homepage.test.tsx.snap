// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomepageContainer /> renders 1`] = `
<div
  class="col"
>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <strong>
        Car
      </strong>
       Management
    </div>
    <div
      class="card-body"
    >
      <form
        action=""
        class="form-horizontal"
        method="post"
      >
        <div
          class="required row mb-3"
        >
          <div
            class="col-md-3"
          >
            <label
              class="form-label"
              for="hf-name required"
            >
              Name
            </label>
          </div>
          <div
            class="col-12 col-md-9"
          >
            <input
              class="is-invalid form-control"
              id="hf-name"
              name="name"
              placeholder="Enter Car Name..."
              type="text"
              value=""
            />
            <small
              class="help-block form-text text-muted"
            >
              Please enter car name
            </small>
          </div>
        </div>
        <div
          class="required row mb-3"
        >
          <div
            class="col-md-3"
          >
            <label
              class="form-label"
              for="hf-brand"
            >
              Brand
            </label>
          </div>
          <div
            class="col-12 col-md-9"
          >
            <select
              class="form-control is-invalid form-select"
              id="hf-brand"
              name="brand"
            >
              <option
                value=""
              >
                Please select
              </option>
              <option
                value="PORSCHE"
              >
                PORSCHE
              </option>
              <option
                value="BMW"
              >
                BMW
              </option>
              <option
                value="AUDI"
              >
                AUDI
              </option>
            </select>
            <small
              class="help-block form-text text-muted"
            >
              Please enter car brand
            </small>
          </div>
        </div>
        <div
          class="required row mb-3"
        >
          <div
            class="col-md-3"
          >
            <label
              class="form-label"
              for="hf-drive"
            >
              Drive
            </label>
          </div>
          <div
            class="required col-12 col-md-9"
          >
            <div
              class="is-invalid form-control col-md-9"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input form-check-input"
                  id="inline-2WD"
                  name="drive2wd"
                  type="checkbox"
                />
                <label
                  class="form-check-label form-check-label"
                  for="inline-2WD"
                >
                  2WD
                </label>
              </div>
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input form-check-input"
                  id="inline-4WD"
                  name="drive4wd"
                  type="checkbox"
                />
                <label
                  class="form-check-label form-check-label"
                  for="inline-4WD"
                >
                  4WD
                </label>
              </div>
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input form-check-input"
                  id="inline-AWD"
                  name="driveawd"
                  type="checkbox"
                />
                <label
                  class="form-check-label form-check-label"
                  for="inline-AWD"
                >
                  AWD
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="required row mb-3"
        >
          <div
            class="col-md-3"
          >
            <label
              class="form-label"
              for="hf-price"
            >
              Price
            </label>
          </div>
          <div
            class="required col-12 col-md-9"
          >
            <input
              class="is-invalid form-control"
              id="hf-price"
              name="price"
              placeholder="Enter Price..."
              type="text"
              value=""
            />
            <small
              class="help-block form-text text-muted"
            >
              Please enter car price.
              <strong
                class="ml-2"
              >
                Note
              </strong>
              : only allow input numbers
            </small>
          </div>
        </div>
        <div
          class="row mb-3"
        >
          <div
            class="col-md-3"
          >
            <label
              class="form-label"
              for="hf-image"
            >
              Image Url
            </label>
          </div>
          <div
            class="col-12 col-md-9"
          >
            <input
              class="form-control"
              id="hf-image"
              name="image"
              placeholder="Enter Image Url..."
              type="text"
              value=""
            />
            <small
              class="help-block form-text text-muted"
            >
              Please enter image url.
              <strong
                class="ml-2"
              >
                Note
              </strong>
              : only allow web image url
            </small>
          </div>
        </div>
      </form>
      <div>
        <div
          class="alert alert-primary fade"
          role="alert"
        >
          test
        </div>
      </div>
    </div>
    <div
      class="card-footer"
    >
      <button
        class="btn btn-primary btn-sm"
        data-testid="add-car-btn"
        type="submit"
      >
        <i
          class="fa fa-dot-circle-o"
        />
         Add Car
      </button>
      <button
        class="offset-sm-1 btn btn-primary btn-sm disabled"
        data-testid="update-car-btn"
        disabled=""
        type="submit"
      >
        <i
          class="fa fa-dot-circle-o"
        />
         Update Car
      </button>
      <button
        class="offset-sm-1 float-right btn btn-danger btn-sm"
        data-testid="reset-car-btn"
        type="submit"
      >
        <i
          class="fa fa-dot-circle-o"
        />
         Reset
      </button>
    </div>
  </div>
</div>
`;
